x-pack/plugins/security_solution/server/lib/detection_engine/signals/source_fields_merging/strategies/merge_all_fields_with_source.ts

```diff
@@ -0,0 +1,113 @@
+    (merged, [fieldsKey, fieldsValue]: [string, FieldsType]) => {
+      if (
+        hasEarlyReturnConditions({
+          fieldsValue,
+          fieldsKey,
+          merged,
+        })
+      ) {
+        return merged;
+      }
+
+      const valueInMergedDocument = get(fieldsKey, merged);
+      if (valueInMergedDocument === undefined) {
+        const valueToMerge = recursiveUnboxingFields(fieldsValue, valueInMergedDocument);
+        return set(fieldsKey, valueToMerge, merged);
+      } else if (isPrimitive(valueInMergedDocument)) {
+        const valueToMerge = recursiveUnboxingFields(fieldsValue, valueInMergedDocument);
+        return set(fieldsKey, valueToMerge, merged);
+      } else if (isArrayOfPrimitives(valueInMergedDocument)) {
+        const valueToMerge = recursiveUnboxingFields(fieldsValue, valueInMergedDocument);
+        return set(fieldsKey, valueToMerge, merged);
+      } else if (
+        isObjectLikeOrArrayOfObjectLikes(valueInMergedDocument) &&
+        isNestedObject(fieldsValue) &&
+        !Array.isArray(valueInMergedDocument)
+      ) {
+        const valueToMerge = recursiveUnboxingFields(fieldsValue, valueInMergedDocument);
+        return set(fieldsKey, valueToMerge, merged);
+      } else if (
+        isObjectLikeOrArrayOfObjectLikes(valueInMergedDocument) &&
+        isNestedObject(fieldsValue) &&
+        Array.isArray(valueInMergedDocument)
+      ) {
+        const valueToMerge = recursiveUnboxingFields(fieldsValue, valueInMergedDocument);
+        return set(fieldsKey, valueToMerge, merged);
+      } else {
+        // fail safe catch all condition for production, but we shouldn't try to reach here and
+        // instead write tests if we encounter this situation.
+        return merged;
+      }
+    },
+    { ...source }
+  );
+
+  return {
+    ...doc,
+    _source: transformedSource,
+  };
+};
+
+/**
+ * Returns true if any early return conditions are met which are
+ *   - If the fieldsValue is an empty array return
+ *   - If we have an array within the path return and the value in our merged documented is non-existent
+ *   - If the value is an object or is an array of object types and we don't have a nested field
+ * @param fieldsValue The field value to check
+ * @param fieldsKey The key of the field we are checking
+ * @param merged The merge document which is what we are testing conditions against
+ * @returns true if we should return early, otherwise false
+ */
+const hasEarlyReturnConditions = ({
+  fieldsValue,
+  fieldsKey,
+  merged,
+}: {
+  fieldsValue: FieldsType;
+  fieldsKey: string;
+  merged: SignalSource;
+}) => {
+  const valueInMergedDocument = get(fieldsKey, merged);
+  return (
+    fieldsValue.length === 0 ||
+    (valueInMergedDocument === undefined && arrayInPathExists(fieldsKey, merged)) ||
+    (isObjectLikeOrArrayOfObjectLikes(valueInMergedDocument) &&
+      !isNestedObject(fieldsValue) &&
+      !isTypeObject(fieldsValue))
```

Comment:
Maybe a more descriptive name for this function -- `typeObject` being an object returned within `fields` that has a `type` key (i.e. geopoint).

Reply:
As discussed this is hard to rename and we are going to leave it as is.